File: serve.py
"""Local static server. Prefers Node — Python's http.server wedges on Windows + Chrome."""
from __future__ import annotations

import os
import shutil
import subprocess
import sys
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer

PORT = int(os.environ.get("PORT", "3456"))
ROOT = os.path.dirname(os.path.abspath(__file__))


def run_node() -> int | None:
    node = shutil.which("node")
    script = os.path.join(ROOT, "serve.mjs")
    if not node or not os.path.isfile(script):
        return None
    env = os.environ.copy()
    env["PORT"] = str(PORT)
    return subprocess.call([node, script], cwd=ROOT, env=env)


class Handler(SimpleHTTPRequestHandler):
    protocol_version = "HTTP/1.0"

    def guess_type(self, path: str) -> str:
        if path.endswith(".wasm.br"):
            return "application/wasm"
        if path.endswith(".js.br"):
            return "application/javascript"
        if path.endswith(".data.br"):
            return "application/octet-stream"
        if path.endswith(".wasm"):
            return "application/wasm"
        return super().guess_type(path)

    def end_headers(self) -> None:
        self.send_header("Connection", "close")
        clean = self.path.split("?", 1)[0]
        if clean.endswith(".br"):
            self.send_header("Content-Encoding", "br")
        if clean.endswith((".html", ".js", ".css")) or clean.endswith("/") or "SaveBruge" in clean:
            self.send_header("Cache-Control", "no-store, max-age=0")
        super().end_headers()

    def log_message(self, format: str, *args) -> None:
        if "favicon.ico" in str(args[:1]):
            return
        super().log_message(format, *args)


class Server(ThreadingHTTPServer):
    allow_reuse_address = os.name != "nt"
    daemon_threads = True
    block_on_close = False
    request_queue_size = 128


if __name__ == "__main__":
    os.chdir(ROOT)
    code = run_node()
    if code is not None:
        raise SystemExit(code)
    with Server(("127.0.0.1", PORT), Handler) as httpd:
        print(f"Serving portfolio at http://127.0.0.1:{PORT}", flush=True)
        print("Unity .br files are served with Content-Encoding: br", flush=True)
        print("Press Ctrl+C to stop.", flush=True)
        httpd.serve_forever()

File: serve.mjs
import http from 'http';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const ROOT = path.dirname(fileURLToPath(import.meta.url));
const PORT = Number(process.env.PORT || 3456);
const HOST = '127.0.0.1';

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8',
  '.mjs': 'application/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.webp': 'image/webp',
  '.ico': 'image/x-icon',
  '.mp4': 'video/mp4',
  '.webm': 'video/webm',
  '.wasm': 'application/wasm',
  '.txt': 'text/plain; charset=utf-8',
  '.woff2': 'font/woff2',
  '.woff': 'font/woff',
};

function typeFor(file) {
  if (file.endsWith('.wasm.br')) return 'application/wasm';
  if (file.endsWith('.js.br')) return 'application/javascript';
  if (file.endsWith('.data.br')) return 'application/octet-stream';
  return MIME[path.extname(file).toLowerCase()] || 'application/octet-stream';
}

function safePath(urlPath) {
  const clean = decodeURIComponent(urlPath.split('?')[0].split('#')[0]);
  const rel = clean === '/' ? '/index.html' : clean;
  const resolved = path.normalize(path.join(ROOT, rel));
  if (!resolved.startsWith(ROOT)) return null;
  return resolved;
}

const server = http.createServer((req, res) => {
  req.socket.setTimeout(0);
  try {
    let file = safePath(req.url || '/');
    if (!file) {
      res.writeHead(400).end();
      return;
    }
    fs.stat(file, (err, st) => {
      if (err || !st.isFile()) {
        if (!err && st.isDirectory()) {
          file = path.join(file, 'index.html');
        } else {
          res.writeHead(404).end('Not found');
          return;
        }
      }
      fs.stat(file, (err2, st2) => {
        if (err2 || !st2.isFile()) {
          res.writeHead(404).end('Not found');
          return;
        }
        const headers = {
          'Content-Type': typeFor(file),
          'Content-Length': st2.size,
          Connection: 'close',
        };
        if (file.endsWith('.br')) headers['Content-Encoding'] = 'br';
        if (/\.(html|js|css)$/i.test(file) || file.includes('SaveBruge')) {
          headers['Cache-Control'] = 'no-store, max-age=0';
        }
        res.writeHead(200, headers);
        const stream = fs.createReadStream(file);
        stream.on('error', () => {
          try { res.destroy(); } catch { /* ignore */ }
        });
        req.on('aborted', () => stream.destroy());
        stream.pipe(res);
      });
    });
  } catch {
    try { res.writeHead(500).end(); } catch { /* ignore */ }
  }
});

server.keepAliveTimeout = 0;
server.headersTimeout = 10_000;
server.requestTimeout = 0;
server.maxConnections = 128;

server.listen(PORT, HOST, () => {
  console.log(`Serving portfolio at http://${HOST}:${PORT}`);
  console.log('Unity .br files are served with Content-Encoding: br');
  console.log('Press Ctrl+C to stop.');
});
